python/mcap-ros2-support/mcap_ros2/cdr.py

```diff
@@ -0,0 +1,162 @@
+"""Decoding of CDR (Common Data Representation) data."""
+
+import struct
+from enum import IntEnum
+
+
+class EncapsulationKind(IntEnum):
+    """Represents the kind of encapsulation used in a CDR stream."""
+
+    CDR_BE = 0  # Big-endian
+    CDR_LE = 1  # Little-endian
+    PL_CDR_BE = 2  # Parameter list in big-endian
+    PL_CDR_LE = 3  # Parameter list in little-endian
+
+
+class CdrReader:
+    """Parses values from CDR data."""
+
+    __slots__ = ("data", "offset", "little_endian")
+
+    def __init__(self, data: bytes):
+        if len(data) < 4:
+            raise ValueError(
+                f"Invalid CDR data size {len(data)}, must contain at least a 4-byte header"
+            )
+        kind = struct.unpack_from("B", data, 1)[0]
+        self.data = data
+        self.offset = 4
+        self.little_endian = kind & 1 == 1
+
+    def kind(self) -> EncapsulationKind:
+        return struct.unpack_from("B", self.data, 1)[0]
+
+    def decoded_bytes(self) -> int:
+        return self.offset
+
+    def byte_length(self) -> int:
+        return len(self.data)
+
+    def int8(self) -> int:
+        value = struct.unpack_from("b", self.data, self.offset)[0]
+        self.offset += 1
+        return value
+
+    def uint8(self) -> int:
+        value = struct.unpack_from("B", self.data, self.offset)[0]
+        self.offset += 1
+        return value
+
+    def int16(self) -> int:
+        self.__align(2)
+        fmt = "<h" if self.little_endian else ">h"
+        value = struct.unpack_from(fmt, self.data, self.offset)[0]
+        self.offset += 2
+        return value
+
+    def uint16(self) -> int:
+        self.__align(2)
+        fmt = "<H" if self.little_endian else ">H"
+        value = struct.unpack_from(fmt, self.data, self.offset)[0]
+        self.offset += 2
+        return value
+
+    def int32(self) -> int:
+        self.__align(4)
```

Comment:
private members/methods only need one underscore before their name, two implies that you want to invoke [name mangling](https://en.wikipedia.org/wiki/Name_mangling), which I don't see a use for here.

Reply:
why not default to double-underscore for all private members? (by the way, I think we use a lot of `__` prefixing in the core mcap python lib)

Reply:
It wouldn't break anything, it's just not idiomatic - unless we're using name mangling for something, let's not tell the reader that we need it.

Reply:
Also re: other libraries, I think we should clean that up at some point.

Reply:
I guess I never learned to read this as “name mangling is _needed_”. Rather I just read it as “this variable does not (and cannot) need to be accessed anywhere outside the class definition” (much like new JS #private identifiers). Are there style guides that discuss this somewhere? I don’t feel strongly about changing it or not, just want to understand if I should change my thinking.

Reply:
["Generally, double leading underscores should be used only to avoid name conflicts with attributes in classes designed to be subclassed."](https://peps.python.org/pep-0008/#method-names-and-instance-variables)